# Transfer Learning MNIST

* Train a simple convnet on the MNIST dataset the first 5 digits [0..4].
* Freeze convolutional layers and fine-tune dense layers for the classification of digits [5..9].

### Import MNIST data and create 2 datasets with one dataset having digits from 0 to 4 and other from 5 to 9 

#### Import the mnist dataset from keras datasets

In [17]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [49]:
import keras
from keras.datasets import cifar10, mnist
from keras.models import Sequential
from keras.layers import Dense, Activation, Dropout, Flatten, Reshape
from keras.layers import Conv2D, MaxPooling2D
from keras.utils import np_utils


In [2]:
from keras.datasets import mnist

Using TensorFlow backend.


In [3]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11493376/11490434 [==============================] - 11s 1us/step


#### Creating two datasets one with digits below 5 and one with 5 and above

In [27]:
x_train_lt5 = x_train[y_train < 5]
y_train_lt5 = y_train[y_train < 5]
x_test_lt5 = x_test[y_test < 5]
y_test_lt5 = y_test[y_test < 5]

x_train_gt5 = x_train[y_train >= 5]
y_train_gt5 = y_train[y_train >= 5] - 5  # make classes start at 0 for
x_test_gt5 = x_test[y_test >= 5]         # np_utils.to_categorical
y_test_gt5 = y_test[y_test >= 5] - 5

### Check 

Verify shapes of x_train, y_train, x_test and y_test for both the datasets with the below given shapes.

In [28]:
print(x_train_lt5.shape)
print(y_train_lt5.shape)
print(x_test_lt5.shape)
print(y_test_lt5.shape)

(30596, 28, 28)
(30596,)
(5139, 28, 28)
(5139,)


In [29]:
print(x_train_gt5.shape)
print(y_train_gt5.shape)
print(x_test_gt5.shape)
print(y_test_gt5.shape)

(29404, 28, 28)
(29404,)
(4861, 28, 28)
(4861,)


### Let us take only the dataset (x_train, y_train, x_test, y_test) for Integers 0 to 4 in MNIST
### Reshape x_train and x_test to a 4 Dimensional array (channel = 1) to pass it into a Conv2D layer

In [30]:
x_train_lt5 = x_train_lt5.reshape(x_train_lt5.shape[0], 28, 28, 1)
x_train_gt5 = x_train_gt5.reshape(x_train_gt5.shape[0], 28, 28, 1)

In [31]:
x_test_lt5 = x_test_lt5.reshape(x_test_lt5.shape[0], 28, 28, 1)
x_test_gt5 = x_test_gt5.reshape(x_test_gt5.shape[0], 28, 28, 1)

### Change into float32 datatype and Normalize x_train and x_test by dividing it by 255.0

In [32]:
x_train_lt5 = x_train_lt5.astype('float32')
x_train_gt5 = x_train_gt5.astype('float32')
x_test_lt5 = x_test_lt5.astype('float32')
x_test_gt5 = x_test_gt5.astype('float32')


In [33]:


x_train_lt5/=255
x_train_gt5/=255
x_test_lt5/=255
x_test_lt5/=255

### Check

Verify the shapes of the X_train and X_test with the shapes given below.

In [34]:
print('x_train_lt5 shape:', x_train_lt5.shape)
print('x_test_lt5 shape:', x_test_lt5.shape)

x_train_lt5 shape: (30596, 28, 28, 1)
x_test_lt5 shape: (5139, 28, 28, 1)


### Use One-hot encoding to divide y_train and y_test into required no of output classes

In [35]:
from keras.utils import np_utils

number_of_classes = 5

y_train_lt5 = np_utils.to_categorical(y_train_lt5, number_of_classes)
y_test_lt5 = np_utils.to_categorical(y_test_lt5, number_of_classes)

y_train_gt5 = np_utils.to_categorical(y_train_gt5, number_of_classes)
y_test_gt5 = np_utils.to_categorical(y_test_gt5, number_of_classes)

In [36]:
y_train_lt5.shape
y_test_lt5.shape
y_train_gt5.shape
y_test_gt5.shape

(30596, 5)

(5139, 5)

(29404, 5)

(4861, 5)

### Build a sequential model with 2 Convolutional layers with 32 kernels of size (3,3) followed by a Max pooling layer of size (2,2) followed by a drop out layer to be trained for classification of digits 0-4  

In [66]:
from keras import backend as K
K.clear_session()

In [67]:


# Define Model
    model3 = Sequential()

    # 1st Conv Layer
    model3.add(Conv2D(filters=32, kernel_size=3, input_shape=(28, 28, 1), activation = 'relu'))
   
    # 2nd Conv Layer
    #model3.add(Convolution2D(32, 3, 3))
    #model3.add(Activation('relu'))

    # Max Pooling
    model3.add(MaxPooling2D(pool_size=(2,2)))
    
    # Dropout
    model3.add(Dropout(0.25))

   

### Post that flatten the data and add 2 Dense layers with 128 neurons and neurons = output classes with activation = 'relu' and 'softmax' respectively. Add dropout layer inbetween if necessary  

In [68]:
 # Fully Connected Layer
    model3.add(Flatten())
    model3.add(Dense(128, activation = 'relu' ))
    
    # More Dropout
    model3.add(Dropout(0.5))

    # Prediction Layer
    model3.add(Dense(5, activation = 'softmax' ))
   

    # Loss and Optimizer
    model3.compile(loss='categorical_crossentropy', optimizer='sgd', metrics=['accuracy'])
    

In [69]:
model3.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 13, 13, 32)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 5408)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               692352    
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 5)                

In [70]:

BATCH_SIZE = 32
EPOCHS = 5

In [71]:
    # Store Training Results
    early_stopping = keras.callbacks.EarlyStopping(monitor='val_acc', patience=7, verbose=1, mode='auto')
    callback_list = [early_stopping]

    # Train the model
    model3.fit(x_train_lt5, y_train_lt5, batch_size=BATCH_SIZE, nb_epoch=EPOCHS, 
              validation_data=(x_test_lt5, y_test_lt5), callbacks=callback_list)

C:\Users\ashutosh.c.agarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:7: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  import sys


Train on 30596 samples, validate on 5139 samples
Epoch 1/5
30596/30596 [==============================] - 20s 661us/step - loss: 0.3289 - accuracy: 0.9036 - val_loss: 1.5915 - val_accuracy: 0.2209
Epoch 2/5
  192/30596 [..............................] - ETA: 24s - loss: 0.1553 - accuracy: 0.9479

C:\Users\ashutosh.c.agarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\callbacks\callbacks.py:846: RuntimeWarning: Early stopping conditioned on metric `val_acc` which is not available. Available metrics are: val_loss,val_accuracy,loss,accuracy
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


30596/30596 [==============================] - 23s 745us/step - loss: 0.1462 - accuracy: 0.9559 - val_loss: 1.5798 - val_accuracy: 0.2209
Epoch 3/5
30596/30596 [==============================] - 23s 741us/step - loss: 0.1194 - accuracy: 0.9628 - val_loss: 1.5857 - val_accuracy: 0.2209
Epoch 4/5
30596/30596 [==============================] - 22s 709us/step - loss: 0.1088 - accuracy: 0.9680 - val_loss: 1.5835 - val_accuracy: 0.2812
Epoch 5/5
30596/30596 [==============================] - 23s 761us/step - loss: 0.0976 - accuracy: 0.9694 - val_loss: 1.5792 - val_accuracy: 0.2343


### Print the training and test accuracy for 5 epochs

### training accuracy is 96%
### test accuracy is 23.4%

### Use the model trained on 0 to 4 digit classification and train it on the dataset which has digits 5 to 9  (Using Transfer learning keeping only the dense layers to be trainable)

### Make only the dense layers to be trainable and convolutional layers to be non-trainable

#### Check model summary to see model layer names

In [61]:
model3.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_5 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
dropout_7 (Dropout)          (None, 13, 13, 32)        0         
_________________________________________________________________
flatten_5 (Flatten)          (None, 5408)              0         
_________________________________________________________________
dense_7 (Dense)              (None, 128)               692352    
_________________________________________________________________
dropout_8 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 5)                

C:\Users\ashutosh.c.agarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\engine\training.py:297: UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
  'Discrepancy between trainable weights and collected trainable'


In [59]:
for layers in model3.layers:
    print(layers.name)
    if('dense' not in layers.name):
        layers.trainable = False
        print(layers.name + 'is not trainable\n')
    if('dense' in layers.name):
        print(layers.name + ' is trainable\n')

conv2d_5
conv2d_5is not trainable

max_pooling2d_4
max_pooling2d_4is not trainable

dropout_7
dropout_7is not trainable

flatten_5
flatten_5is not trainable

dense_7
dense_7 is trainable

dropout_8
dropout_8is not trainable

dense_8
dense_8 is trainable



In [62]:
model3.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_5 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
dropout_7 (Dropout)          (None, 13, 13, 32)        0         
_________________________________________________________________
flatten_5 (Flatten)          (None, 5408)              0         
_________________________________________________________________
dense_7 (Dense)              (None, 128)               692352    
_________________________________________________________________
dropout_8 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 5)                

### Do the required preprocessing for `x_train_gt5` also same as `x_train_lt5` and for `y_train_gt5` same as `y_train_lt5`

1. Reshape
2. Change to float32 datatype
3. Normalize (dividing with 255)
4. y_train and y_test Convert into one-hot vectors

In [0]:
## already done earlier

### Check

Verify the shapes with the given below.

In [63]:
print(x_train_gt5.shape)
print(y_train_gt5.shape)
print(x_test_gt5.shape)
print(y_test_gt5.shape)

(29404, 28, 28, 1)
(29404, 5)
(4861, 28, 28, 1)
(4861, 5)


## Print the accuracy for classification of digits 5 to 9

In [64]:

BATCH_SIZE = 32
EPOCHS = 1

In [65]:
    # Store Training Results
    early_stopping = keras.callbacks.EarlyStopping(monitor='val_acc', patience=7, verbose=1, mode='auto')
    callback_list = [early_stopping]

    # Train the model
    model3.fit(x_train_gt5, y_train_gt5, batch_size=BATCH_SIZE, nb_epoch=EPOCHS, 
              validation_data=(x_test_gt5, y_test_gt5), callbacks=callback_list)

Train on 29404 samples, validate on 4861 samples
Epoch 1/1
  224/29404 [..............................] - ETA: 19s - loss: 2.3864 - accuracy: 0.3170

C:\Users\ashutosh.c.agarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:7: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  import sys


29404/29404 [==============================] - 18s 628us/step - loss: 0.4398 - accuracy: 0.8573 - val_loss: 21.1722 - val_accuracy: 0.9455


C:\Users\ashutosh.c.agarwal\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\callbacks\callbacks.py:846: RuntimeWarning: Early stopping conditioned on metric `val_acc` which is not available. Available metrics are: val_loss,val_accuracy,loss,accuracy
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


### Training Accuracy is 85%
### Test Accuracy is 94.5%

# Text classification using TF-IDF

###  Load the dataset from sklearn.datasets

In [133]:
from sklearn.datasets import fetch_20newsgroups

In [134]:
categories = ['alt.atheism', 'soc.religion.christian', 'comp.graphics', 'sci.med']

### Training data

In [135]:
twenty_train = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)

### Test data

In [136]:
twenty_test = fetch_20newsgroups(subset='test', categories=categories, shuffle=True, random_state=42)

###  a.  You can access the values for the target variable using .target attribute 
###  b. You can access the name of the class in the target variable with .target_names


In [137]:
twenty_train.target

array([1, 1, 3, ..., 2, 2, 2], dtype=int64)

In [138]:
twenty_train.target_names

['alt.atheism', 'comp.graphics', 'sci.med', 'soc.religion.christian']

In [139]:
twenty_train.data[0:5]

['From: sd345@city.ac.uk (Michael Collier)\nSubject: Converting images to HP LaserJet III?\nNntp-Posting-Host: hampton\nOrganization: The City University\nLines: 14\n\nDoes anyone know of a good way (standard PC application/PD utility) to\nconvert tif/img/tga files into LaserJet III format.  We would also like to\ndo the same, converting to HPGL (HP plotter) files.\n\nPlease email any response.\n\nIs this the correct group?\n\nThanks in advance.  Michael.\n-- \nMichael Collier (Programmer)                 The Computer Unit,\nEmail: M.P.Collier@uk.ac.city                The City University,\nTel: 071 477-8000 x3769                      London,\nFax: 071 477-8565                            EC1V 0HB.\n',
 "From: ani@ms.uky.edu (Aniruddha B. Deglurkar)\nSubject: help: Splitting a trimming region along a mesh \nOrganization: University Of Kentucky, Dept. of Math Sciences\nLines: 28\n\n\n\n\tHi,\n\n\tI have a problem, I hope some of the 'gurus' can help me solve.\n\n\tBackground of the probl

In [140]:
import pandas as pd
import numpy as np
import scipy as sp

from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

### Now with dependent and independent data available for both train and test datasets, using TfidfVectorizer fit and transform the training data and test data and get the tfidf features for both

In [141]:
# create a document-term matrix using TF-IDF
vect = TfidfVectorizer()
pd.DataFrame(vect.fit_transform(twenty_train).toarray(), columns = vect.get_feature_names())
features_train = vect.get_feature_names()


data  descr  filenames  target  target_names
0   1.0    0.0        0.0     0.0           0.0
1   0.0    0.0        1.0     0.0           0.0
2   0.0    0.0        0.0     0.0           1.0
3   0.0    0.0        0.0     1.0           0.0
4   0.0    1.0        0.0     0.0           0.0

In [142]:
X_train_tfidf = np.array(vect.fit_transform(twenty_train.data).toarray())

In [143]:
y_train_tfidf = twenty_train.target

In [144]:
X_test_tfidf = np.array(vect.fit_transform(twenty_test.data).toarray())

In [145]:
y_test_tfidf = twenty_test.target

In [146]:
X_test_tfidf.shape

(1502, 30786)

In [147]:
X_train_tfidf.shape

(2257, 35788)

In [148]:
# create a document-term matrix using TF-IDF
#vect = TfidfVectorizer(stop_words='english')
#dtm_test = vect.fit_transform(twenty_test.data)
#features_test = vect.get_feature_names()
#dtm_test.shape

### Use logisticRegression with tfidf features as input and targets as output and train the model and report the train and test accuracy score

In [149]:
# use logistic regression with text column only
logreg = LogisticRegression(C=1e9)
logreg.fit(X_train_tfidf, y_train_tfidf)



LogisticRegression(C=1000000000.0, class_weight=None, dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False)

In [150]:
y_pred_class = logreg.predict(X_train_tfidf)

print(metrics.accuracy_score(y_train_tfidf, y_pred_class))

1.0


In [152]:
y_pred_class = logreg.predict(X_test_tfidf)

print(metrics.accuracy_score(y_test_tfidf, y_pred_class))

ValueError: X has 30786 features per sample; expecting 35788